mindmeld/components/entity_resolver.py

```diff
@@ -42,60 +50,277 @@
+                           "See https://www.mindmeld.com/docs/userguide/entity_resolver.html "
+                           "for more details.")
+            er_config = copy.deepcopy(er_config)
+            model_settings = er_config.get("model_settings", {})
+            model_settings.update({"resolver_type": model_type})
+            er_config["model_settings"] = model_settings
+            er_config["model_type"] = "resolver"
 
-    def __init__(
-        self, app_path, resource_loader, entity_type, es_host=None, es_client=None
-    ):
-        """Initializes an entity resolver
+    return er_config
+
+
+class EntityResolverFactory:
+
+    @staticmethod
+    def _validate_resolver_type(name):
+        if name not in ENTITY_RESOLVER_MODEL_TYPES:
+            raise Exception(f"Expected 'resolver_type' in ENTITY_RESOLVER_CONFIG "
+                            f"among {ENTITY_RESOLVER_MODEL_TYPES}")
+        if name == "sbert_cosine_similarity" and not _is_module_available("sentence_transformers"):
+            raise ImportError(
+                "Must install the extra [bert] by running `pip install mindmeld[bert]` "
+                "to use the built in embbedder for entity resolution.")
+
+    @classmethod
+    def create_resolver(cls, app_path, resource_loader, entity_type, **kwargs):
+        """
+        Identifies appropriate entity resolver based on input config and
+            returns it.
 
         Args:
-            app_path (str): The application path
-            resource_loader (ResourceLoader): An object which can load resources for the resolver
-            entity_type: The entity type associated with this entity resolver
-            es_host (str): The Elasticsearch host server
+            app_path (str): The application path.
+            resource_loader (ResourceLoader): An object which can load resources for the resolver.
+            entity_type (str): The entity type associated with this entity resolver.
+            er_config (dict): A classifier config
+            es_host (str): The Elasticsearch host server.
+            es_client (Elasticsearch): The Elasticsearch client.
         """
-        self._app_namespace = get_app_namespace(app_path)
-        self._resource_loader = resource_loader
-        self._normalizer = resource_loader.query_factory.normalize
+        er_config = (
+            kwargs.pop("er_config", None) or
+            get_classifier_config("entity_resolution", app_path=app_path)
+        )
+        er_config = _correct_deprecated_er_config(er_config)
+        resolver_type = er_config["model_settings"]["resolver_type"]
+        cls._validate_resolver_type(resolver_type)
+        return ENTITY_RESOLVER_MODEL_MAPPINGS.get(resolver_type)(
+            app_path, resource_loader, entity_type, er_config, **kwargs
+        )
+
+
+class EntityResolverModelBase(ABC):
+    """
+    Base class for Entity Resolvers
+    """
+
+    def __init__(self, app_path, resource_loader, entity_type, er_config):
```

Comment:
The function should only receive as much information it needs to compute on it. Instead of passing in the `resource_loader ` whose main purpose is to provide access to the `entity_map ` and `normalizer `, its preferable to pass in those objects directly. That way, the function is easier to understand and gets us close to single-use type functions.

Reply:
Thanks for pointing that out. Will make necessary changes.